pkg/configschema/fields_test.go

```diff
@@ -0,0 +1,186 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configschema
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReadFields_Config(t *testing.T) {
+	fields, err := readFields(
+		reflect.ValueOf(testConfig{Name: "foo"}),
+		nopCommentReader,
+	)
+	require.NoError(t, err)
+	assert.Equal(t, "configschema.testConfig", fields.Type)
+	nameField := fields.CfgFields[0]
+	assert.Equal(t, "name", nameField.Name)
+	assert.Equal(t, "string", nameField.Kind)
+	assert.Equal(t, "foo", nameField.Default)
+}
+
+func TestReadFields_TestStructWithDefaults(t *testing.T) {
+	ts := testStruct{
+		One:      "1",
+		Two:      2,
+		Three:    3,
+		Four:     true,
+		Duration: 42,
+		Squashed: testPerson{"squashed"},
+		PersonPtr: &testPerson{
+			Name: "foo",
+		},
+		PersonStruct: testPerson{
+			Name: "bar",
+		},
+	}
+	testReadFields(t, ts, map[string]any{
+		"one":           "1",
+		"two":           int64(2),
+		"three":         uint64(3),
+		"four":          true,
+		"duration":      "42ns",
+		"name":          "squashed",
+		"person_ptr":    "foo",
+		"person_struct": "bar",
+	})
+}
+
+func TestReadFields_TestStructWithoutDefaults(t *testing.T) {
+	testReadFields(t, testStruct{}, map[string]interface{}{
+		"one":           "",
+		"three":         uint64(0),
+		"four":          false,
+		"name":          "",
+		"person_ptr":    "",
+		"person_struct": "",
+	})
+}
+
+func testReadFields(t *testing.T, ts testStruct, expectedDefaults map[string]any) {
+	root, err := readFields(reflect.ValueOf(ts), testCommentReader())
+	require.NoError(t, err)
+	assert.Equal(t, "testStruct comment\n", root.Doc)
+	assert.Equal(t, "configschema.testStruct", root.Type)
+
+	assert.Equal(t, 11, len(root.CfgFields))
+
+	assert.Equal(t, &cfgField{
+		Name:    "one",
+		Kind:    "string",
+		Default: expectedDefaults["one"],
+	}, getFieldByName(root.CfgFields, "one"))
+
+	/*
```

Comment:
Why is this commented out? Was it like this on `cmd/configschema`?

Reply:
No problem at all @mx-psi -- thanks again for your reviews! I agree that this PR is trying to do too much, so I am inclined to close it, and either re-open as a few smaller PRs or perhaps move this functionality out of contrib altogether. I will bring this up at the next sig and see what folks say. Saludos!